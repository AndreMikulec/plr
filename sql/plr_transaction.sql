CREATE TABLE test1 (a int, b text);


CREATE PROCEDURE transaction_test1()
LANGUAGE plr
AS $$
  version_11plus  <- pg.spi.exec("select current_setting('server_version_num')::integer >= 110000;")
  if(version_11plus[[1]])
  {
    for(i in 0:9)
    {
      pg.spi.exec(paste("INSERT INTO test1 (a) VALUES (", i, ")"))
      if (i %% 2 == 0) 
      {
        pg.spi.commit()
      } 
      else 
      {
        pg.spi.rollback()
      }
    }
  }
  else
  {
    pg.spi.exec("INSERT INTO test1 (a) VALUES (0);")
    pg.spi.exec("INSERT INTO test1 (a) VALUES (2);")
    pg.spi.exec("INSERT INTO test1 (a) VALUES (4);")
    pg.spi.exec("INSERT INTO test1 (a) VALUES (8);")
  }
$$;

CALL transaction_test1();

SELECT * FROM test1;


TRUNCATE test1;
