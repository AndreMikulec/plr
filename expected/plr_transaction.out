CREATE TABLE test1 (a int, b text);
CREATE OR REPLACE FUNCTION test_create_procedure_transaction() RETURNS void
  AS
$BODY$
  version_11plus  <- pg.spi.exec("select current_setting('server_version_num')::integer >= 110000;")
  if(version_11plus[[1]])
  {
    pg.spi.exec("
      CREATE OR REPLACE PROCEDURE transaction_test1()
        AS
      $$
        for(i in 0:9){
          pg.spi.exec(paste('INSERT INTO test1 (a) VALUES (', i, ');'))
          if (i %% 2 == 0) {
            pg.spi.commit()
          } else {
            pg.spi.rollback()
          }
        }
      $$ LANGUAGE plr;
      ")
    pg.spi.exec("CALL transaction_test1();")
  }
  else
  {
    pg.spi.exec("INSERT INTO test1 (a) VALUES (0);")
    pg.spi.exec("INSERT INTO test1 (a) VALUES (2);")
    pg.spi.exec("INSERT INTO test1 (a) VALUES (4);")
    pg.spi.exec("INSERT INTO test1 (a) VALUES (8);")
  }
$BODY$
LANGUAGE plr;
SELECT test_create_procedure_transaction();
 test_create_procedure_transaction 
-----------------------------------
 
(1 row)

SELECT * FROM test1;
 a | b 
---+---
 0 | 
 2 | 
 4 | 
 8 | 
(4 rows)

